fix: yespower: do not use abort() with cache PoW

decryp2kanon opened the pull request:
https://github.com/ResistancePlatform/resistance-core/commit/64ee78a1f51c37bd87a2f46c652ff88b0f3f7e2b

https://github.com/sugarchain-project/sugarchain/issues/37

decryp2kanon committed:
fix: yespower exit()

decryp2kanon committed:
cleanup: thash << hash

decryp2kanon committed:
fix: don't use abort() << exit(1)

decryp2kanon committed:
fix: static const

decryp2kanon committed:
(const uint8_t *) << (const unsigned char *)

decryp2kanon committed:
(const uint8_t *) vs (unsigned char *)

cryptozeny commented on src/primitives/block.cpp:
```diff
@@ -27,20 +29,21 @@ uint256 CBlockHeaderUncached::GetHash() const
+    uint256 hash;
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << *this;
+    if (yespower_tls((unsigned char *)&ss[0], ss.size(), &yespower_1_0_sugarchain, (yespower_binary_t *)&hash)) {
```
@solardiz in `yespower_tls`, should we use `(const uint8_t *)` instead of `(unsigned char *)` here? it was unsigned char.

decryp2kanon replied:
https://github.com/sugarchain-project/sugarchain/blob/0624be48d82c38367c108b84c5532ca95e30263f/src/crypto/yespower-1.0.1/yespower-opt.c#L1129-L1130

decryp2kanon committed:
fix: (const uint8_t *) in yespower_tls
```
int yespower_tls(const uint8_t *src, size_t srclen,
    const yespower_params_t *params, yespower_binary_t *dst)
```

decryp2kanon commented:
https://github.com/sugarchain-project/sugarchain/issues/37

decryp2kanon committed:
cleanup

decryp2kanon committed:
fprintf: Uncached vs _cached

decryp2kanon committed:
cleanup

cryptozeny commented on src/primitives/block.cpp:
```diff
@@ -11,55 +11,58 @@
+    uint256 hash;
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << *this;
+    if (yespower_tls((const uint8_t *)&ss[0], ss.size(), &yespower_1_0_sugarchain, (yespower_binary_t *)&hash)) {
```
`const uint8_t` instead of `unsigned char` are you sure?

decryp2kanon replied:
need to check compiler log
https://stackoverflow.com/questions/7009231/cast-unsigned-char-uint8-t-to-const-char

decryp2kanon replied:
no compiler warning
https://github.com/sugarchain-project/sugarchain/issues/63

decryp2kanon commented:
solo mining on testnet is fine.
https://1explorer-testnet.cryptozeny.com/tx/adda5ba73f889500758094cf71694c32c7f64361562088169216d351fc5ef071

cryptozeny reviewed (approved):
Tested ACK 071a57f

volbil commented:
Looking good ACK

decryp2kanon commented:
`-reindex-chainstate` took `5` hrs
`-reindex` took `10` hrs
no problem found.

decryp2kanon commented:
log
```
x block = 3a5580ada582ca819f8f4cc8a9eba73507c5c2322d8659c36183dc3d1e336b71 PoW = 00000098179aa1d2df06a7c98be2f9cfb32cc0925690a1d26e3b38d6b76f1073
x block = 3a5580ada582ca819f8f4cc8a9eba73507c5c2322d8659c36183dc3d1e336b71 PoW = 00000098179aa1d2df06a7c98be2f9cfb32cc0925690a1d26e3b38d6b76f1073
x block = f1436fe38845e2e28b8dc2228b86d9cf88a8ceb66f8f7c5aad4226e3279b4472 PoW = 000000ef8fb2cc233edd41887c980cca76b9e2181bff3499fa563b4e2be7a93c
x block = f1436fe38845e2e28b8dc2228b86d9cf88a8ceb66f8f7c5aad4226e3279b4472 PoW = 000000ef8fb2cc233edd41887c980cca76b9e2181bff3499fa563b4e2be7a93c
x block = 6747149bfd39ed370bf02ce7bc83c63b593d3a76bed4bab0b34d31f5dad80a4a PoW = 00000057f9f39892220827a87f5297c75d5b5d77454c2104cf25756e8d2af6e6
x block = 6747149bfd39ed370bf02ce7bc83c63b593d3a76bed4bab0b34d31f5dad80a4a PoW = 00000057f9f39892220827a87f5297c75d5b5d77454c2104cf25756e8d2af6e6
x block = 1ec76d0b6f55ada3f61f7ae915c2fac41d8f58f8104c62c7e0eafb7a487d72ee PoW = 000000bbd38c69e72e32db51654ca57164b8450956108674a20da589cf764b8d
x block = 1ec76d0b6f55ada3f61f7ae915c2fac41d8f58f8104c62c7e0eafb7a487d72ee PoW = 000000bbd38c69e72e32db51654ca57164b8450956108674a20da589cf764b8d
```

debug
```
2020-02-09 09:38:50 UpdateTip: new best=3a5580ada582ca819f8f4cc8a9eba73507c5c2322d8659c36183dc3d1e336b71 height=2925365 version=0x20000000 log2_work=44.056938 tx=3127481 date='2020-02-09 09:38:50' progress=1.000000 cache=278.6MiB(1729334txo)
2020-02-09 09:39:00 UpdateTip: new best=f1436fe38845e2e28b8dc2228b86d9cf88a8ceb66f8f7c5aad4226e3279b4472 height=2925366 version=0x20000000 log2_work=44.056939 tx=3127482 date='2020-02-09 09:38:50' progress=0.999999 cache=278.6MiB(1729336txo)
2020-02-09 09:39:01 UpdateTip: new best=6747149bfd39ed370bf02ce7bc83c63b593d3a76bed4bab0b34d31f5dad80a4a height=2925367 version=0x20000000 log2_work=44.05694 tx=3127483 date='2020-02-09 09:39:00' progress=1.000000 cache=278.6MiB(1729338txo)
2020-02-09 09:39:02 UpdateTip: new best=1ec76d0b6f55ada3f61f7ae915c2fac41d8f58f8104c62c7e0eafb7a487d72ee height=2925368 version=0x20000000 log2_work=44.056941 tx=3127484 date='2020-02-09 09:39:01' progress=1.000000 cache=278.6MiB(1729340txo)
```

decryp2kanon commented:
#61

cryptozeny commented on src/primitives/block.cpp:
```diff
@@ -11,55 +11,58 @@
     if (cache_init) {
         if (block_hash != cache_block_hash) {
-            fprintf(stderr, "Error: CBlockHeader: block hash changed unexpectedly\n");
+            fprintf(stderr, "Error: GetPoWHash_cached: block hash changed unexpectedly\n");
```
should `CBlockHeader_cached`

decryp2kanon replied:
https://github.com/sugarchain-project/sugarchain/pull/69

decryp2kanon commented:
https://github.com/cryptozeny/sugarchain-v0.16.3/issues/120
